### ISSUE TITLE
[d3] Strategy for D3 Version 5 Definitions Release [BREAKING D3 version 4?]

### ISSUE THREAD

**Issue by tomwanzek** [DefinitelyTyped/DefinitelyTyped#23257]
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @tomwanzek @gustavderdrache @Ledragon 

The release cycle for D3 Version 5 has just started with a first Release Candidate:
* **d3-request** will be dropped in its entirety and replaced with **d3-fetch**,
* **d3-contour** will be added, and
* there is a major re-shuffle including feature deprecations between **d3-scale** and **d3-scale-chromatic** combined with the addition of **d3-scale-chromatic** to the bundle

In particular, the feature deprecation and reshuffle with regard to **d3-scale** and **d3-scale-chromatic** (see [release notes](https://github.com/d3/d3/releases/tag/v5.0.0)), imply that we need to pin down the minor versions for the latest/last D3 v4 bundle definitions.

I.e. we need to add a `package.json` stub back in, so we do not break users running v4, before we add the new bundle definitions for D3 version 5.

@gustavderdrache and @Ledragon any comments/thoughts re preparing for D3 v5?

@andy-ms Any thoughtful guidance, how we ensure we can safely support the migration between version 4 and version 5 definitions without breaking users of v4?

Cheers, T

Related to #23250, #23251 and #23253.

#23252 (d3-selection 1.3) needs to be merged into the final definitions bundle for d3: 4.13.x

**Comment by Ledragon** [comment]
Dang, you beat me to it!
I think we need to start by implementing d3-fetch (as per #23253), and update d3-scale, and d3-scale-chromatic.

As you mentioned, we need to ensure backwards compatibility. I have no experience in managing dependencies between typings packages, so I don't know how this could be achieved

**Comment by tomwanzek** [comment]
@Ledragon if you tackle `d3-fetch`, I will tackle `d3-select` (v1.3). I can bring `d3-select` into the current `d3` v 4.x definitions release cycle. I will check, if bringing a `package.json` stub s.a.:

```
{
   private: true,
   dependencies: {
     // ...
     "@types/d3-scale": "1.0"
    // ...
   }
}
````
will do the trick. It used to 😄 Then we would use the pinning of dependencies to reasonable upper/lower bounds to control the cut-over from D3 v4 to D3v5

**Comment by Ledragon** [comment]
Just a shower thought: do you think we could sync with Mike Bostock to include each type directly with each module? This would ensure versions consistency, but could slow down the release cycle, so unsure how to feel about it.

**Comment by tomwanzek** [comment]
The question has been raised with Mike before. There does not seem to be sufficient interest on his part to engage on the matter. His call, his priorities/preferences,  I suppose.

[Question + Offer without response](https://github.com/d3/d3/issues/2855#issuecomment-257745074)
And in the case of d3-sankey a clear no: d3/d3-sankey#15.

So, as for my part, I am always open to have the dialog about pros and cons to make an informed decision. Mike, with all his otherworldly, appreciated efforts, seems to be focused on other matters.

D3 bundle ~1.2 million downloads last month, D3 bundle TS definitions ~185,000. So, about 15% of "new installs". Solid, but hey, whatcha gonna do? :smile:

**Comment by Ledragon** [comment]
I guess we could try getting back on the subject by offering to maintain the definitions ourselves in each individual repo, making it clear that it would not become an overhead for him. From my point of view, I don't really care about submitting PRs there or here. On the other hand, I can understand why Mike would not want to endorse the responsibility of shipping the types with the packages, so I don't really know...

**Comment by tomwanzek** [comment]
@andy-ms We have a material issue with BREAKING CHANGES related to the D3 definitions. And need some urgent guidance from someone on the DefinitelyTyped tooling side.

As you know, the D3 Bundle definitions in `d3` are re-exporting the definitions of the constituent bundles. As it appears, the ability of pin down the `@types/d3-***` dependencies using a `package.json` stub in the `d3` folder has been removed without suitable replacement.

Users of the version 4.x definitions for `d3` **will break (!!!)**, as soon as we publish the latest version of the definitions for `d3-scale` (major version 2) and `d3-scale-chromatic` (minor version 1.3). Including installed base in production. These modules have been materially refactored:
* features from `d3-scale` have been migrated to `d3-scale-chromatic`
* a feature of `d3-scale` has been removed without replacement (it was previously marked as deprecated but still included)
* in D3 version 4 `d3-scale-chromatic` is not part of the standard `d3` bundle. Given the refactoring, it becomes part of the D3 version 5 standard bundle.

We also cannot force them to just upgrade to D3 version 5, as there are other material breaking changes, i.e. some constituent modules are getting completely replaced.

So thanks in advance for a quick response, or looping the right person into this thread, if it isn't you this time around!!!

**Comment by andy-ms** [comment]
Are you planning to update the d3-* packages all at once, or one at a time?
If all at once, you could add `package.json`-s now to pin down the version to v4 (this will also require making a pull request to [types-publisher](https://github.com/Microsoft/types-publisher/blob/master/dependenciesWhitelist.txt) to allow dependencies on these), and then make a pull request updating everything to v5 together. 
If one at a time, we do support the ability to have both v4 and v5 at the same time, and you could use path mapping to point to older versions of a package. See the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-update-a-package-to-a-new-major-version). This would probably be more total work than an all-at-once update.

**Comment by tomwanzek** [comment]
@andy-ms Thanks for getting back on this so quickly. What I had in mind, was the following:

(1) Add a `package.json` stub to the current `d3` minor version 4.12 to lock down its `@types`-dependencies to minor versions. And publish the resulting patch release.
(2) Once the `d3-selection` minor version 1.3 PR #23323 is merged, Update the `d3` to minor version 4.13 and pin its `d3-selection` dependency to `1.3`.

The resulting `d3` version 4.13.0 would then be the final expected release published for D3 version 4. However, it would continue to benefit from any _patch_ releases to the constituent module definitions. On the other hand, it should be protected against breaking retroactively due to _minor/major_ changes like the `d3-scale`/`d3-scale-chromatic` swap.

After(1) and (2) are done, we can start:
(3) publishing otherwise retroactively D3v4-breaking changes like `d3-scale` v2.0.0 and `d3-scale-chromatic` v1.2.0.
(4) Update the `d3` Definitions to version 5.0 by updating its dependencies (module swaps and new major/minor versions of constituent modules.)

I have tried pushing (1) already to my DefinitelyTyped fork already for a  PR, but the commit failed on Travis (see https://github.com/tomwanzek/DefinitelyTyped/commit/1302dd09ff1fd459fe4bc956874e445203766590 and the [Travis Log](https://travis-ci.org/tomwanzek/DefinitelyTyped/builds/335820503?utm_source=github_status&utm_medium=notification))

I will have to double-check if any other D3-modules have dependencies on `d3-scale`, we would have to then proceed correspondingly.

**Q1** If I understand you correctly, after whitelisting `d3` in **types-publisher**, the above commit on my fork should work and we can merge the PR? If so, then I will submit a PR as per your proposal.

**Q2** I have to digest the parallel maintenance implications for "historic" major definitions in a subfolder (i.e. `v1` for `d3-scale`, `v4` for `d3`) a little more, but the approach seems good. Just to confirm  though: Could we then fix, say, a definitions error in the historic `v1` folder for `d3-scale` and it would publish to `npm@types` under a 1.x.y new patch release?

As always, incredibly indebted for the amazing support 💯 

@gustavderdrache @Ledragon Please chime in, if I overlooked anything or am going down the wrong garden path...

**Comment by tomwanzek** [comment]
I just check all the `d3-xxx` modules @gustavderdrache, @Ledragon and/or myself are maintaining. None of them has a _direct_ dependency on `d3-scale`. `d3-array`, `d3-axis` and `d3-zoom` use `d3-scale` in their test-file only. However, the imports will _not_ be affected by the refactoring of `d3-scale` v2 and `d3-scale-chromatic` v1.2.

So we do not have to pin those at his point.

I cannot vouch for any other packages in the DT repo.

**Comment by gustavderdrache** [comment]
We're doing parallel maintenance already. The d3v3 typings are still in the repo [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3/v3/index.d.ts) and were updated pretty recently. You can see the result on unpkg (note the version specifier): https://unpkg.com/@types/d3@3/index.d.ts.

It seems like we'd still have to have a package.json to lock down dependency versions (since the v4 bundle is just an amalgamation of other packages), but otherwise keeping a parallel v4 shouldn't cause too many issues.

**Comment by andy-ms** [comment]
If a `v4` folder exists you should be able to use a path mapping to that instead of using `package.json`. Path mapping is preferred since we don't have to do any `npm install` to test the package.

**Comment by andy-ms** [comment]
> protected against breaking retroactively due to minor/major changes like the d3-scale/d3-scale-chromatic swap.

Important: the mechanism we use doesn't protect against accidentally upgrading a package's major version and forgetting to add path mappings for its dependers, so you'll have to be careful to remember to do that when updating a package.

**Comment by tomwanzek** [comment]
But how will the published versions _remain_ pinned, if we only use path mapping in the v4 folder? Also, path-mapping alone does not address the issue, since `d3-scale-chromatic` will experience only a minor version update from 1.1 to 1.2. But `d3-scale-chromatic` v1.2 is only meaningful in a bundle with `d3-scale` v2. Otherwise they contain some duplicate possibly conflicting features.

**Comment by tomwanzek** [comment]
At the end of the day, we need to outcomes:
* DT can run all compilations/tests correctly, **and**
* the published `package.json` is continually pinned to the right major/minor dependencies rather than `*`.

**Comment by andy-ms** [comment]
> But how will the published versions remain pinned, if we only use path mapping in the v4 folder?

You can have a path mapping from the latest version of a package to an older version of one of its dependencies. So for every dependency, you get to choose whether you're depending on the latest version or on an older version.

**Comment by tomwanzek** [comment]
**TLDR;** I understand that path-mapping in `tsconfig.json` is more efficient for the testing infrastructure. However, I fail to see how it ensures that the `package.json` files published on npm for `@types/d3` will have the bare minimum dependency constraints imposed such that the `d3` **4.12+** definitions do not break user installs. (For **<4.12** it's already too late!!!)

Let me be explicit about the requirements as I see them. Below I will use **U-#** to indicate what I consider to be **user requirements** and **T-#** for what I consider to be **technical requirements**.

## Current State

The `@types/d3` bundle definitions are published in minor release version **4.12.x**. Once PR #23323 (`d3-selection`  new minor v1.3.0) is merged, we can create a new release for `@types/d3` with version **4.13.0**. This would correspond to the _currently_ last version for D3, i.e. **v4.13.0**

The actual D3 **4.13.0** dependencies are pinned as follows (focusing on the ones that matter to illustrate the issue):

```
"dependencies": {
    "d3-request": "1.0.6",
    "d3-scale": "1.0.7",
    // other "d3-xxx":  "x.y.z" dependencies
  }
```
_Outside_ of the `d3` bundle for version 4.x, a standalone module `d3-scale-chromatic` is maintained. `d3-scale-chromatic` **1.1** peacefully coexists with `d3-scale`  **1.x**.

With the release of `d3-scale` **2.0** several features where removed. _All except one_ have been added to the latest minor version of `d3-scale-chromatic` **1.2**

As of D3 **5.0.0-rc.2** the actual D3 dependencies will be:

```
"dependencies": {
    "d3-contour": "1", // New addition
    "d3-fetch": "1", // wholesale  replacement for "d3-request"
    "d3-scale": "2", // New major with removed features
    "d3-scale-chromatic": "1", // New addition (part of refactoring with d3-scale v2)
    // other "d3-xxx":  "x" dependencies
  }
```
Note: That on **5+** basis, `d3` will be pinned only to major release versions. See [D3 Issue comment](https://github.com/d3/d3/issues/3256#issuecomment-361250655)

**_Assuming we make no changes to how we pin dependencies_** for `d3` and publish definitions for:
* `d3-scale` **2.0** (new major)
* `d3-scale-chromatic` **1.2** (new minor), the following will happen:

The _published_ `package.json` file for `@types/d3` **4.13** and all previous **4.x** releases follow the default DT/types-publisher behaviour and _do not_ pin dependencies. Specifically, they (will) have been  published including the following dependency:

```
  "dependencies": {
    "@types/d3-scale": "*",
    // other "@types/d3-xxx":  "*" dependencies
  }
```
Bar any pre-existing yarn/npm lock files, installs of `d3` **4.x** with `@types/d3` **4.x** will break because they will pull `@types/d3-scale` **2.x** which had key features removed and is not compatible with  `d3` **4.x**

## Expected Behaviour

### (U-1) D3 Version 4 Install Compatibility

`@types/d3` **4.12** and **4.13** **_MUST be published_** with `package.json` dependencies including:
```
  "dependencies": {
    "@types/d3-scale": "1.0",
    // other "@types/d3-xxx":  "x.y" dependencies
  }
```
**_IMPORTANT_**: Users with `d3` and `@types/d3`  **<4.12** will still be broken unless their installs are npm/yarn locked locally. So they will have to either upgrade or manually pin their `@types/d3` sub-module dependencies.

### (U-2) D3 Version 5 Install Compatibility

After publishing `@types/d3` **4.12** and **4.13** as per **(U-1)** and publishing `@types/d3-scale` **2.0.0** and `@types/d3-scale-chromatic` **1.2**, `@types/d3` **5.x** **_MUST be published_** with `package.json` dependencies including:
```
  "dependencies": {
    "@types/d3-contour": "1",
    "@types/d3-fetch": "1",
    "@types/d3-scale": "2",
    "@types/d3-scale-chromatic": "1",
    // other "@types/d3-xxx":  "x" dependencies
   // Removed "@types/d3-request"
  }
```
These are pinned to major versions of the `d3-xxx` modules in accordance with the pinning policy of `d3`  **5+** itself.

This is required, so we we do not run into the current issue again in the future, i.e. future major releases of `d3` or `d3-xxx` modules.

### (U-3) D3 Version 4 Parallel Support

On a best efforts basis, we **MUST** be able to continue supporting the historic `d3` **4.x** bundle. Which means, possibly updating and publishing a new `@types/d3` **4.x** minor release we updated `package.json` dependencies. (Pinned as per **(U-1)**)

### (T-1) Correct Dependencies in DefinitelyTyped Compile/Lint Tests 

DT scripts and CI tests must use the correct, applicable version of `d3-xxx` definitions when testing dependent package definitions (here: `d3`, for which we do not really have tests, as its only a barrel).

### (T-2) "Efficient" DT Test Scripts/CI Environment Usage

I am not even attempt to formalize this piece here, as you will be more familiar with the details than I could possibly be. I do appreciate the speed gains in the tests script from early DT to now 💯 🎉 
Clearly, this motivated the preference for avoiding/minimizing npm installs and favouring path-mapping.

## Preliminary Solution Assessment

So, finally :wink: with all of the above itemized...

As far as I see, `tsconfig.json` path-mapping addresses the technical requirements **(T-1)** and **(T-2)**.
With the limitation, that path-mapping works only in corresponding to _major_ releases maintained in **\v#** history version folders. It cannot distinguish files at minor version level, since we simply edit the respective `index.ts` and test files in the major version/current folder.

**(U-3)** Parallel support is _partially_ addressed in that we can create a **\v4** history folder for `d3` and a **\v1** history folder for **d3-scale**.

I understand that, with path-mapping we could map from **\v4** of `d3` to **\v1** of **d3-scale**. By implication, DT tests (local and CI) would run in accordance with "quasi-pinned" major versions.

However, I fail to see how it addresses the fundamental **published** `package.json` requirements described under user requirements **(U-1)**, **(U-2)** and **(U-3)**.


Thanks in advance for clearing things up for me, in case I am still on the wrong garden path 😄 
I still seem to be leaning towards Microsoft/types-publisher#432 .

**Comment by tomwanzek** [comment]
@andy-ms Any update on the above so we can proceed to a change, where we can publish pinned dependencies for `d3` **4.12/4.13** before proceeding to the actual migration to `d3` **5**?

This is a blocking issue. Thanks in advance.

**Comment by andy-ms** [comment]
Talked with @mhegazy and not convinced we need more than the current functionality to support an update here.

U-1) Why is it essential to pin to a *minor* rather than *major* version? Is it just to avoid including *too much* in the type definition, since the minor version update could add new functionality?

U-2) Similarly, is this requirement simply to avoid publishing v5 with "*" dependencies? It's been our experience that the same library will rarely make major type-breaking changes even when updating a major version, so the default has been to use "*" instead of a major version so that when a major version changes, we don't have to update all dependers to depend on the new major version.

**Comment by tomwanzek** [comment]
> It's been our experience that the same library will rarely make major type-breaking changes even when updating a major version, so the default has been to use "" 

Clearly, the changes between D3 **4.13** and D3 **5.0** discussed in **Current State** [above](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/23257#issuecomment-362477088) constitute backward breaking changes introduced by `d3-scale` **1.x** -> **2.x** for all `@types/d3` **4.x** definitions without dependency constraints. We are not talking hypotheticals here. `@types/d3` is downloaded some ~200K/month.

User requirement **(U-2)**, the pinning of `d3` **5** dependencies, is geared at preventing us from, again, ending up in the same predicament as now.

> Why is it essential to pin to a minor rather than major version? Is it just to avoid including too much in the type definition, since the minor version update could add new functionality?

For one, absolutely, the mental model of TS is that, if my code compiles, it will run without runtime errors. So, yes, relying on a TS definitions API that exposes properties/methods which do not exist in the underlying JS module goes against the very grain of this core user affordance of TS.

We have a long track record of keeping the ~30 D3 module definitions current. In the beginning, before the `package.json`-stub was removed from `d3`, we kept it current by pinning to minor versions routinely. The effort was close to zero and the frequency of the changes very tolerable.

> not convinced we need more than the current functionality to support an update here.

**A question for added clarity:** If we use the **_same_** DT PR to
* update **d3-scale** to version **2**,
* move the old **d3-scale** version **1** definitions to a subfolder **\v1**
* apply `tsconfig.json` path-mapping in the `d3` **4.12** definition to **d3-scale\v1**

Will the resulting new **published version** of `@types\d3` **4.12.x** have a `package.json` dependency constraint of:
```
{
    dependencies: {
     // ...
     "@types/d3-scale": "1"
    // ...
   }
}
```
or will it have been published with "@types/d3-scale": "*" again?

I.e. does types-publisher parse the path-mapping and translate them into major version dependency constraints  in the `package.json`?

**Comment by andy-ms** [comment]
A path-mapping introduces a major version dependency, since you're explicitly relying on an older version of the package instead of the newest. A dependency on the latest version of a package will have a "*" dependency.

If the update is made all at once, users of `@types/d3@4` who upgrade the the latest v4 release will get a new version whose dependencies are pinned to major versions. Users who instead upgrade to `@types/d3@5` will  have dependencies on "*", which presumably will also upgrade to the latest versions of those packages.

We don't supported minor version folders (e.g. `v1.2`) on DefinitelyTyped, so there would be no way to maintain separate `d3-scale@1.2` and `d3-scale@1.3` types. It's true that this means the types a user ends up downloading may contain more functions than their older installed packages do, but we'll have to see if this results in major user complaints.

**Comment by tomwanzek** [comment]
Well then, let's see how it goes.

**Step 1) Close out Prerequisites**

* **d3-fetch** definitions 2nd iteration PR #23378 is ready to merge by DT maintainer
* **d3-selection** minor version 1.3 in PR #23323 needs approval and merge (@gustavderdrache )

**Step 2) Refactor D3 v4/v5 and Update d3-scale and d3-scale-chromatic**

Once Step 1) is complete, I will create a _**single PR**_ which will:
* Move **d3-scale** 1.x into a `v1` subfolder
* Update **d3-scale** to 2.0.0
* Update **d3-scale-chromatic** to minor version 1.2
* Move D3 v4 into a subfolder `v4` while
   - updating the version number to `4.13`
   - apply the path-mapping to `d3-scale\v1`
* Update D3 latest to version 5.0, while
   - updating the imports/re-exports to reflect the new composition (**d3-fetch** in, **d3-request** out, **d3-scale-chromatic** in, **d3-contour** in)

**Comment by tomwanzek** [comment]
Getting started on the PR in accordance with Step 2).

**Comment by tomwanzek** [comment]
Slight change of plans: I just noticed, that the **d3-scale-chromatic** definitions were never updated from 1.0.2 to 1.1.x. For continuity, if will push a PR to bring **d3-scale-chromatic up to 1.1.1 for publication, before conducting the refactoring as per Step 2.

**Comment by tomwanzek** [comment]
My bad, correction: **d3-scale-chromatic** has been publish in minor version 1.x, I had just parsed my usual reference line for the last state of affairs:
```ts
// Last module patch version validated against: 1.0.2
```
which still referred to an older version.

Back on track 😄
